### ISSUE TITLE
tf.keras.layers.BatchNormalization API does not fully support masking

### ISSUE THREAD

**Issue by LukeBolly** [tensorflow/tensorflow#30021]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): TF 1.13.1
- Python version: 3.6.7
- CUDA/cuDNN version: 
- GPU model and memory:

**Describe the current behavior**
Trying to apply a mask when calling a BatchNormalization layer fails with:
*TypeError: call() got an unexpected keyword argument 'mask'*

**Describe the expected behavior**
Masking should be supported:
https://github.com/tensorflow/tensorflow/blob/93dd14dce2e8751bcaab0a0eb363d55eb0cc5813/tensorflow/python/keras/layers/normalization.py#L188

If a layer supports masking, it should be a kwarg to __call__. Until all layers support masking, it's not practical to only support masking via an upstream layer. A Masking Layer can't always be inserted into a model because downstream convolution layers don't support it at all. Feeding the mask into the call of a recurrent layer is the only workaround to combine Convolution, LSTM and masking, and it would be great if it worked the same way for BatchNorm:
```encoded = keras.layers.LSTM(10)(input, mask=input_mask)```.

**Code to reproduce the issue**
```
from tensorflow import keras

def Works():
    input = keras.layers.Input(shape=(None, 1))
    masked_input = keras.layers.Masking(-1)(input)
    normalized = keras.layers.BatchNormalization(epsilon=1.1e-5)(masked_input)

def DoesntWork():
    input = keras.layers.Input(shape=(None, 1))
    mask = keras.layers.Masking(-1).compute_mask(input)
    normalized = keras.layers.BatchNormalization(epsilon=1.1e-5)(input, mask=mask)

if __name__ == '__main__':
    Works()
```

**Comment by achandraa** [comment]
In order to expedite the trouble-shooting process, please provide a full code snippet to reproduce the issue reported here. Thanks!

**Comment by LukeBolly** [comment]
If you run this, you will see the error:

```
from tensorflow import keras

def DoesntWork():
    input = keras.layers.Input(shape=(None, 1))
    mask = keras.layers.Masking(-1).compute_mask(input)
    normalized = keras.layers.BatchNormalization(epsilon=1.1e-5)(input, mask=mask)

if __name__ == '__main__':
    DoesntWork()
```
*TypeError: call() got an unexpected keyword argument 'mask'*

**Comment by achandraa** [comment]
I have tried on Colab with TF version 1.13.1 and was able to reproduce issue.

**Comment by LukeBolly** [comment]
Is the Mask actually used when calculating the mean/variance? The padded values shouldn't be included in the calculations or they are going to bias the mean and lower the variance.

**Comment by LukeBolly** [comment]
This is probably part of a bigger issue: #30161
If the layer incorrectly uses the Mask (ie. ignores it) and causes a model to silently fail, then it should do what every other layer that doesn't support masking does and throw an error when you try and input a Mask. The current behavior is even worse than getting an error because it implies that it will operate as expected, when it is most likely impacting performance.

**Comment by tanzhenyu** [comment]
The way masking works is that we categorize all layers into three categories:
1) producer, that has compute_mask
2) consumer, that takes `mask` inside call()
3) some kind of passenger, that simply pass through the masking.

for 3), masking is "silently" supported by self.support_masking=True and you don't have to pass `mask` into call(), because we treat batch norm as passing through the mask, not `consuming` the mask. On the other hand, recurrent layers are consumer of mask, i.e., passing `mask` into the call makes sure the output reflects it. recurrent layers will also produce mask as well. Mask layer is merely a producer.

That said, in terms of masking, batch norm is just like dense or conv or activation layers, they all fall into 3).

**Comment by LukeBolly** [comment]
In the case of Convolutional layers, that's not true.

```
from tensorflow import keras

if __name__ == '__main__':
    input = keras.layers.Input(shape=(None, None, 1))
    masked_input = keras.layers.Masking(-1)(input)
    conv = keras.layers.Conv2D(1,3)(masked_input)
    normalized = keras.layers.BatchNormalization(epsilon=1.1e-5)(conv)
```
*TypeError: Layer conv2d does not support masking, but was passed an input_mask: Tensor("masking/Any_1:0", shape=(?, ?, ?), dtype=bool)*

I'd argue that BatchNorm (and probably Dense layers too if they have bias nodes) also don't support masking because they don't work as expected with it. Conv2d could pass the mask through too, but instead the user is notified that it won't work which is much better behavior in my opinion. At least in my case, I knew I needed to work the mask around my conv layers. I did not know the same thing for BatchNormalization.

Is the logic that:
1. If a layer has self.support_masking = True, then it passes it through.
2. If a layer also has a `mask` argument in call, then it consumes the mask.
3. If a layer has neither, it will throw an error because it cannot pass the mask and cannot consume it.

To be fair I should have looked further than `self.support_masking=True`, but to me it seems that there is either an issue with consistency or with documentation. This is the first time I've seen those 3 categories of Masking you've listed above;
